### DISCUSSION TITLE
Rendering Composition Animations - IBrush changes across Key Frames

### DISCUSSION THREAD

**Discussion by Nolan5G** [AvaloniaUI/Avalonia#16270] [Q&A]
Is there support with CompositionVisuals to essentially create a ColorKeyFrameAnimation without having to go the CustomVisual route?

Context:
So I was having a look at amwx's Fluent Avalonia sample splash screen, and felt inspired to really see what was possible with the CompositionVisuals.

I came upon issues when I grabbed one for a Path element, and a runtime error couldn't find Fill amongst the property list for the composition visual.

So I delved further into Avalonia and found CompositionPropertySet at the heart of determining what properties are available and providing it.  It seems to support only value properties.

I'm sorta confused on what the design spec was for the composition renderer, and where it's limitations are toward these kind of animated visual compositions.

**Comment by maxkatz6** [comment]
> I'm sorta confused on what the design spec was for the composition renderer, and where it's limitations are toward these kind of animated visual compositions.

Composition animations were added as a side-feature for the new composition renderer that was implemented for 11.0 release. While renderer itself was a main and stable feature, animations were shipped as an experimental API.

It's not prioritized right now, and more APIs, including ColorKeyFrameAnimation, still has to be implemented. 
General idea is to follow UWP-like composition APIs - but by itself UWP API doesn't have a detailed spec, making it rather difficult.

**Reply by stevemonaco** [reply]
Windows Community Toolkit ships an [AnimationBuilder](https://learn.microsoft.com/en-us/windows/communitytoolkit/animations/animationbuilder) for UWP/WinUI3 created by @Sergio0694 that should cover the surface area necessary for app developers to create high quality animations. Of course, this isn't the underlying compositor implementation.

**Reply by maxkatz6** [reply]
@stevemonaco yeah, I previously ported [ExpressionBuilder](https://learn.microsoft.com/en-us/windows/communitytoolkit/animations/expressions) https://github.com/AvaloniaUI/Avalonia.Labs/tree/main/src/Avalonia.Labs.ExpressionBuilder. AnimationBuilder seems to be an addition on top of that.
